### PR TITLE
Avro Schema registry kafka pubsub implementation

### PR DESCRIPTION
**Pull request by passuied**
# Description

Add Avro Schema Registry  to Kafka PubSub.
- Leverages [srclient](https://github.com/riferrei/srclient) and [go-avro](https://github.com/linkedin/goavro) under the hood
- Please note that for the serialization/deserialization to fully compatible, dates/datetimes in the JSON data need to be set as their respective Unix int and long representations, instead of the more standard Iso8601 one. This is a limitation of the goavro library.

## Issue reference

#3144

# Checklist

 - [x] Code compiles correctly
 - [x] Created/updated tests
 - [x] Extended the documentation - Created issue

### CODE REVIEW

**Comment by passuied** [comment]
Please let me know if it's preferable to add publisher in same PR or separate.

**Comment by yaron2** [comment]
Great contribution, I really like how this is done and wonder if this approach can be used in the future to enable a generic schema validation feature for all Dapr pub/subs, implemented in the Dapr runtime itself.

For this PR: It'd be best to add a certification test for this functionality (or add this to an existing test).

**Comment by passuied** [comment]
> For this PR: It'd be best to add a certification test for this functionality (or add this to an existing test).
 Thanks @yaron2, Is there doc about these tests and especially how to run them locally?

**Comment by ItalyPaleAle** [comment]
The certification tests for Kafka PubSub are here: https://github.com/dapr/components-contrib/tree/main/tests/certification/pubsub/kafka

We don't have good docs (we should....), but for certifications test you just "cd" into the test's folder, and then run `go test -v -tags cerrttests`

**Comment by yaron2** [comment]
@passuied please resolve the conflict

File: pubsub/kafka/metadata.yaml [hunk 1]
```diff
@@ -259,4 +259,31 @@ metadata:
...
+      description: |
+        Enables caching for schemas.
+      example: '"true"'
+      default: '"true"'
+    - name: LatestSchemaCacheTTL
+      type: duration
+      description: |
+        The TTL for schema caching when publishing a message with latest schema available.
+      example: '"5min"'
+      default: '"5min"'
```
Comment (ItalyPaleAle) [inline pubsub/kafka/metadata.yaml:289]: Incorrect format for Go durations:
```suggestion
      example: '"5m"'
      default: '"5m"'
```

File: bindings/kafka/metadata.yaml [hunk 2]
```diff
@@ -268,4 +268,31 @@ metadata:
...
+    description: |
+      Enables caching for schemas.
+    example: '"true"'
+    default: '"true"'
+  - name: LatestSchemaCacheTTL
+    type: duration
+    description: |
+      The TTL for schema caching when publishing a message with latest schema available.
+    example: '"5min"'
+    default: '"5min"'
```
Comment (ItalyPaleAle) [inline bindings/kafka/metadata.yaml:298]: Incorrect format for Go durations:
```suggestion
      example: '"5m"'
      default: '"5m"'
```

File: common/component/kafka/kafka.go [hunk 3]
```diff
@@ -55,6 +67,39 @@ type Kafka struct {
...
+)
+
+type SchemaCacheEntry struct {
+	schema         *srclient.Schema
+	codec          *goavro.Codec
+	expirationTime time.Time
+}
+
+func GetValueSchemaType(metadata map[string]string) (SchemaType, error) {
+	schemaTypeStr, ok := metadata[valueSchemaType]
```
Comment (ItalyPaleAle) [inline common/component/kafka/kafka.go:84]: Please use this method from dapr/kit which also supports case-insensitive matching: https://github.com/dapr/kit/blob/77f7f031c92afb648d24d6dea18cf80e07fe21d5/metadata/utils.go#L26

File: common/component/kafka/kafka.go [hunk 4]
```diff
@@ -166,6 +223,128 @@ func (k *Kafka) Close() (err error) {
 	return err
 }
 
+func getSchemaSubject(topic string) string {
+	// For now assumes that subject is named after topic (e.g. `my-topic-value`)
+	return fmt.Sprintf("%s-value", topic)
```
Comment (ItalyPaleAle) [inline common/component/kafka/kafka.go:228]:
```suggestion
	return topic + "-value"
```

File: common/component/kafka/kafka.go [hunk 5]
```diff
@@ -166,6 +223,128 @@ func (k *Kafka) Close() (err error) {
...
+}
+
+func (k *Kafka) DeserializeValue(message *sarama.ConsumerMessage, config SubscriptionHandlerConfig) ([]byte, error) {
+	if config.ValueSchemaType == Avro {
```
Comment (ItalyPaleAle) [inline common/component/kafka/kafka.go:232]: Nit: please use a switch statement here, it just reads better :) 

Alternatively, change this to be short-circuited, such as :

```go
if config.ValueSchemaType != Avro {
    return message.Value, nil
}
```

File: common/component/kafka/kafka.go [hunk 6]
```diff
@@ -166,6 +223,128 @@ func (k *Kafka) Close() (err error) {
...
+
+func (k *Kafka) getLatestSchema(topic string) (*srclient.Schema, *goavro.Codec, error) {
+	srClient, err := k.getSchemaRegistyClient()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	subject := getSchemaSubject(topic)
+	if k.schemaCachingEnabled {
+		cacheEntry, ok := k.latestSchemaCache[subject]
```
Comment (ItalyPaleAle) [inline common/component/kafka/kafka.go:274]: This needs a lock around it or it's not safe for concurrent access

File: common/component/kafka/kafka.go [hunk 7]
```diff
@@ -166,6 +223,128 @@ func (k *Kafka) Close() (err error) {
...
+		}
+		schema, errSchema := srClient.GetLatestSchema(subject)
+		if errSchema != nil {
+			return nil, nil, errSchema
+		}
+		// New JSON standard serialization/Deserialization is not integrated in srclient yet.
+		// Since standard json is passed from dapr, it is needed.
+		codec, errCodec := goavro.NewCodecForStandardJSONFull(schema.Schema())
+		if errCodec != nil {
+			return nil, nil, err
```
Comment (ItalyPaleAle) [inline common/component/kafka/kafka.go:287]:
```suggestion
			return nil, nil, errCodec
```

File: common/component/kafka/kafka.go [hunk 8]
```diff
@@ -166,6 +223,128 @@ func (k *Kafka) Close() (err error) {
...
+	srClient, err := k.getSchemaRegistyClient()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	subject := getSchemaSubject(topic)
+	if k.schemaCachingEnabled {
+		cacheEntry, ok := k.latestSchemaCache[subject]
+		// Cache present and not expired
+		if ok && cacheEntry.expirationTime.Compare(time.Now()) < 0 {
```
Comment (ItalyPaleAle) [inline common/component/kafka/kafka.go:276]: Nit: please use [`time.Before`](https://pkg.go.dev/time#Time.Before) (or `time.After`) for readability

File: common/component/kafka/kafka.go [hunk 9]
```diff
@@ -166,6 +223,128 @@ func (k *Kafka) Close() (err error) {
...
+	return k.srClient, nil
+}
+
+func (k *Kafka) SerializeValue(topic string, data []byte, metadata map[string]string) ([]byte, error) {
+	valueSchemaType, err := GetValueSchemaType(metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	if valueSchemaType == Avro {
```
Comment (ItalyPaleAle) [inline common/component/kafka/kafka.go:320]: Nit: please use a switch statement for readability

File: common/component/kafka/kafka.go [hunk 10]
```diff
@@ -166,6 +223,128 @@ func (k *Kafka) Close() (err error) {
...
+		}
+
+		valueBytes, err := codec.BinaryFromNative(nil, native)
+		if err != nil {
+			return nil, err
+		}
+		schemaIDBytes := make([]byte, 4)
+		binary.BigEndian.PutUint32(schemaIDBytes, uint32(schema.ID()))
+
+		var recordValue []byte
```
Comment (ItalyPaleAle) [inline common/component/kafka/kafka.go:338]: Nit: optimize memory allocation (hopefully I got the right count)

```suggestion
		recordValue := make([]byte, 0, len(schemaIDBytes)+len(valueBytes)+1)
```

File: common/component/kafka/metadata.go [hunk 11]
```diff
@@ -87,6 +88,12 @@ type KafkaMetadata struct {
 	channelBufferSize    int    `mapstructure:"-"`
 	consumerFetchMin     int32  `mapstructure:"-"`
 	consumerFetchDefault int32  `mapstructure:"-"`
+	// schema registry
+	SchemaRegistryURL       string        `mapstructure:"schemaRegistryURL"`
+	SchemaRegistryAPIKey    string        `mapstructure:"schemaRegistryAPIKey"`
+	SchemaRegistryAPISecret string        `mapstructure:"schemaRegistryAPISecret"`
+	SchemaCachingEnabled    bool          `mapstructure:"schemaCachingEnabled"`
+	LatestSchemaCacheTTL    time.Duration `mapstructure:"LatestSchemaCacheTTL"`
```
Comment (ItalyPaleAle) [inline common/component/kafka/metadata.go:96]: I think I'd prefer if this started with "schema" for consistency with the other schema-related properties

```suggestion
	LatestSchemaCacheTTL    time.Duration `mapstructure:"latestSchemaCacheTTL"`
```

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Thanks for this PR :) left some comments

File: common/component/kafka/kafka.go (hunk 7 above)
Comment (passuied) [reply common/component/kafka/kafka.go:287]: Thanks! Great catch!

**Review by ItalyPaleAle — APPROVED** [review]

**Comment by ItalyPaleAle** [comment]
LGTM! @yaron2 good to merge if you are good too

**Comment by yaron2** [comment]
Thanks for this @passuied

**Comment by passuied** [comment]
> Thanks for this @passuied

Thank you for a very smooth PR process ;) That was fun!

**Pull request merged**